tooling/nargo_fmt/src/visitor/item.rs

```diff
@@ -109,6 +112,94 @@ impl super::FmtVisitor<'_> {
         (result.trim_end().to_string(), last_line_contains_single_line_comment(maybe_comment))
     }
 
+    // This formats the function outer doc comments, attributes, modifiers, and `fn name`.
+    fn format_fn_header(&self, src: &str) -> String {
+        let mut result = String::new();
+        let lexer = Lexer::new(src).skip_comments(false);
+        let mut parser = Parser::for_lexer(lexer);
+
+        // First there might be outer doc comments
+        while let Some(token) = parser.eat_kind(TokenKind::OuterDocComment) {
+            result.push_str(&token.to_string());
+            result.push('\n');
+            result.push_str(&self.indent.to_string());
+        }
+
+        // Then, optionally, attributes
+        while let Some((attribute, _span)) = parser.parse_attribute() {
```

Comment:
Why are we using the parser in the formatter here, can we not achieve the same with the lexer alone?
What happens if the formatter receives invalid input here, does it remove it?

Reply:
I tried to use the lexer in the past (a couple of weeks ago). It's slightly trickier because if we get `pub` we have to check for `(`... but I think I bumped into other issues when I tried it. I'll try again now :-)

If there's invalid input then the formatter doesn't run.

Reply:
I managed to do it :-)